doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+  * When the question template is re-rendered, the `error` state variable is used to lookup the appropriate validation error message in the [question template](doc/question-templates.md#-error_message-message-).
+
+> The use of these blocks is encouraged. However, they should call `valid_xxx?` methods on the `calculator` state variable and not rely on the `response` argument passed into the block.
+
+##### `next_node(&block)`
+
+* There must only be one of these blocks per question definition.
+* This block is intended to determine which node comes next based on the user responses so far.
+* These blocks are executed after all the `validate` blocks have been executed and before any `save_input_as` blocks are executed.
+* The built-in state variables, `path`, `current_node` & `responses`, are updated if this block returns successfully.
+* The block return value must be the result of calling the `#question` or `#outcome` methods passing in the key of the next node - see the [next node documentation](doc/next-node-rules.md) for more details.
+
+> The use of this block type is *required*. However, it should call methods on the `calculator` state variable and not rely on the `response` argument passed into the block.
+
+##### `save_input_as(variable_name)`
+
+* Although you can call this method multiple times within a single question, only the last call will actually have any effect.
+* This method is intended to allow you to store the response to the current question for use in blocks in *subsequent* nodes.
+* The response is stored after the `next_node` block has been executed and before any `calculate` blocks are executed.
+* The response is stored on the state object as a state variable named `variable_name`.
+
+> The use of these blocks is deprecated and should never be necessary; `on_response` blocks should be used instead.
+
+##### `calculate(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed in subsequent questions or outcomes.
+* These blocks are executed after `save_input_as` has executed and before any of the `precalculate` blocks on the *next* node are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+#### Further information
+
+See the [documentation on storing data](doc/storing-data.md).
+
+#### Templates
+
+See the [documentation for question templates](doc/question-templates.md).
+
+### Outcome nodes
+
+These are very similar to question nodes, however, only the following methods are available within the node definition:
+
+* [`precalculate`](#precalculatevariable_name-block)
+* [`on_response`](#on_responseblock)
```

Comment:
You can use `on_response` in an outcome? How? I understood that on_response stores user input?